pkg/mutation/system.go

```diff
@@ -0,0 +1,165 @@
+package mutation
+
+import (
+	"fmt"
+	"sort"
+	"sync"
+
+	"github.com/google/go-cmp/cmp"
+	"github.com/open-policy-agent/gatekeeper/pkg/mutation/mutator"
+	"github.com/open-policy-agent/gatekeeper/pkg/mutation/schema"
+	"github.com/pkg/errors"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+)
+
+// System keeps the list of mutations and
+// provides an interface to apply mutations.
+type System struct {
+	schemaDB schema.DB
+	mutators []mutator.Mutator
+	sync.RWMutex
+}
+
+// NewSystem initializes an empty mutation system
+func NewSystem() *System {
+	return &System{
+		mutators: make([]mutator.Mutator, 0),
+	}
+}
+
+// Upsert updates or insert the given object, and returns
+// an error in case of conflicts
+func (s *System) Upsert(m mutator.Mutator) error {
+	s.Lock()
+	defer s.Unlock()
+
+	// Checking schema consistence only if the mutator has schema
+	if withSchema, ok := m.(mutator.WithSchema); ok {
+		err := s.schemaDB.Upsert(withSchema)
+		if err != nil {
+			return errors.Wrapf(err, "Schema upsert failed")
+		}
+	}
+
+	i := sort.Search(len(s.mutators), func(i int) bool {
+		return greaterOrEqual(s.mutators[i], m)
+	})
+
+	if i == len(s.mutators) { // Adding to the bottom of the list
```

Comment:
We should make sure we are storing pointers to mutators, that those pointers are referencing a deepcopy of the mutator (guess we'll need to add `DeepCopy()` to the mutator interface).

Then we can store both a sorted list and a map of mutators. This will allow constant-time lookups for things like checking-for-diffs.

Doing this will make things like repeated reconciliations cheaper.

Reply:
Not sure I got the suggestion right, but replacing the mutators slice with something like

```mutators []*Mutator```

makes the code a bit less easy to read. 

On the other hand, having pointers to the objects (like *Assign) implementing Mutator sounds like a good idea.
Will leave it with out the pointers slice / map, we can get back to this later.

Reply:
Yeah, no need to have a pointer to an interface (which is redundant IIRC). Making the pointer implement the interface is definitely the way to go.

We'll also need DeepCopy to guard against the mutators being edited by the caller.